wordpress-mobile/WordPress-Android
# Introduce GutenbergEditPostActivity

This PR introduces a dedicated Activity for hosting the Gutenberg editor, as opposed to having all editors co-live within `EditPostActivity`, which already has lots of code to make a difference and tell which editor to launch in the first place, depending on user settings, Post contents, and more intricate conditions.

#### Context and goals
It is important to note where this particular effort comes from. This is not a real refactor of `EditPostActivity` which we've been discussing for long and postponing for when time is available to tackle such a non-trivial, and also not minor, effort. 

This PR on the contrary, represents what I've found to be the less disruptive way to achieve these goals:
1. have the Activity hosting Gutenberg not be re-created on orientation changes (although yes on other configuration changes) to let React Native handle that flow (see https://github.com/wordpress-mobile/WordPress-Android/pull/8800). This in particular would take care of fixing these two problems: https://github.com/wordpress-mobile/WordPress-Android/issues/8739 and https://github.com/wordpress-mobile/gutenberg-mobile/issues/377
2. avoid introducing more bugs into the existing editor (particularly Aztec, Visual, and Legacy), that we won't have time to maintain, by isolating any further changes in the time those editors have to still live together.

For this, we're introducing a new Activity to host the Gutenberg editor _only_, but will keep the current `EditPostActivity`.
I've elevated the shared constants and _only_ the most obvious common code into an abstract class, but I've no intention to maximize code reuse for this particular effort. This is a temporal situation that makes Aztec/Visual editors live without much of a problem while we introduce the new Gutenberg editor.
A real, thorough refactor would evidently demand that we not only remove code that is not used, but we'd also need to refactor and reuse code as much as possible. Such an effort would demand more valuable time and thought for a thing that we know we're going to remove soon as part of our known plans. Thus, containing new things in an isolated way and building up from there seems to come in very handy.

#### ToDo (general)
- [x] for new intents, build a helper method that returns the `Intent` we need depending on the Editor we need to launch. (https://github.com/wordpress-mobile/WordPress-Android/pull/8874/commits/13d3d75576f22e8974802b1b406ddf95771cc8ab and https://github.com/wordpress-mobile/WordPress-Android/pull/8874/commits/2ccd1791c6d37d92fd2e40d7649154d11f090c4d)
- [x] get all important constants up into an `abstract EditPostBaseActivity` class, and make `EditPostActivity` a derived class. (https://github.com/wordpress-mobile/WordPress-Android/pull/8874/commits/74765be3b1a5615b2e5a1eb68bc26921aded4a81)
- [x] make a separate GutenbergEditPostActivity that also derives from `EditPostBaseActivity`, and add the configchanges attributes in `AndroidManifest.xml` (https://github.com/wordpress-mobile/WordPress-Android/pull/8874/commits/79eef3230539b86147fd30ba03e016b6e0123e9e)
- [x] check the different entry points everywhere in the app for which we need to decide to open which editor, ~possibly make that logic centralized~ ended up having it all in `ActivityLauncher` where Activity intents where created. (https://github.com/wordpress-mobile/WordPress-Android/pull/8874/commits/2ccd1791c6d37d92fd2e40d7649154d11f090c4d and https://github.com/wordpress-mobile/WordPress-Android/pull/8874/commits/2f8ef0e4d1f2679a336d742580404802c89d1d90)
- [x] once we make sure all entry points work as expected, remove non-Gutenberg related code from `GutenbergEditPostActivity` _(left only commented code 5e1c762148c874c711e293e9b6a98c2088b08c2f)_
- [x] remove Gutenberg code from `EditPostActivity` (see ToDo lists below).

#### ToDo on `GutenbergEditPostActivity`
- [x] remove `mShowGutenbergEditor`  flag and usages as it's unneeded in this context (d008ed1642884521a0e2a20bdeb257178f898943)
- [x] remove `mShowAztec` flag and usages as it won't belong here (5e1c762148)
- [x] remove Aztec references and AztecImageLoader and AztecVideoLoader (5e1c762148)
- [x] remove `mShowNewEditor` flag and usages as it won't belong here (85263bb194)
- [ ] ~make "discardLocalChanges" functionality work the same on Gutenberg as it was on Aztec~ I'll tackle this on a separate PR once this one lands


#### Media - related ToDo (on separate PRs)
(5e1c762148c874c711e293e9b6a98c2088b08c2f commented out code that we'll need to reuse for Media soon)
Moved to issue https://github.com/wordpress-mobile/gutenberg-mobile/issues/459

#### ToDo on `EditPostActivity` (cleanup)
- [x] remove Gutenberg code from `EditPostActivity` (fd181f266d)

#### To test:

CASE A: use gutenberg
1. go to app settings and enable gutenberg (compile the alpha version for this)
2. go to posts list and open a Gutenberg post written on the web
3. see it loads Gutenberg (you should be able to see the blocks interface when tapping somewhere in the body)

CASE B: use Aztec in compatibility mode
1. go to app settings and disable gutenberg (compile the alpha version for this)
2. go to posts list and tap on a Gutenberg post written on the web
3. you should see the compatibility warning dialog, tap OPEN
4. see Aztec is the editor opened, and the content is rendered (to the extent that Aztec can render Gutenberg posts ofc)

CASE C: use Gutenberg and rotate
1. follow case A
2. when the post is open, scroll down to the middle of it (make sure you have several blocks in your post to be able to tell the difference in scroll positioning).
3. tap on a pargraph block and see the keyboard appears, and the caret starts blinking where you tapped.
4. tap back to dismiss the keyboard
5. rotate the device
6. observe the place where you were standing is still there, and the caret is blinking


CASE D: use Gutenberg and rotate, observe undo/redo
1. follow case A
2. when the post is open, scroll down to the middle of it (make sure you have several blocks in your post to be able to tell the difference in scroll positioning).
3. write some new blocks, see the UNDO action becomes available
4. tap UNDO, observe the last action is undone
5. tap REDO, observe the text come back as it was
6. rotate the device
7. observe the UNDO action is still available
8. repeat steps 4 and 5 to confirm that the history was effectively available and actions can be undone/redone.

**NOTE:** other entry points, such as uploading an image to the Media section of your site and then tapping on "WRITE POST" on the success snackbar/notification to include the just-uploaded picture in your post is not yet supported. We'll support that once Media is all wired up together.


Update release notes:
